### ISSUE TITLE
spago build for voidlinux

### ISSUE THREAD

**Issue by waynevanson** [purescript/spago#686]
Hi Guys,

I've been trying to create a pre built binary for void linux as part of the xbps system.
This means that users using voidlinux or the X Binary Package System (XBPS) don't need to install stack or GHC, because it's all been done by the host (cloud) in advance. 

It's critical that these binaries are compiled from source by the host so it can be run on all available architectures without any issues.

When building using the recommended settings by the voidlinux community (`build_style=haskell-stack`), I get the message below:

```log
basement > /tmp/stack-48a6ffdab601d322/basement-0.0.8/Basement/Block/Base.hs:398:9: error:
basement >     Not in scope: ‘fail’
basement >    Perhaps you meant ‘Data.List.tail’ (imported from Data.List)
basement >     |
basement > 398 |         arr@(Block arrBa) <- makeTrampoline
basement >     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
basement >
basement > /tmp/stack-48a6ffdab601d322/basement-0.0.8/Basement/Block/Base.hs:481:9: error:
basement >     Not in scope: ‘fail’
basement >     Perhaps you meant ‘Data.List.tail’ (imported from Data.List)
basement >     |
basement > 481 |         b@(Block ba) <- unsafeFreeze pinnedMb
basement >     |  
```

From what I've read, this happens because `RebindingSyntax` is declared in the file somewhere, therefore `NoImplicitPrelude` applies.

I can get this building on my machine with `stack install` and some dependencies, but not using the required setup. 

I don't use haskell, so everything I know so far is solely off building purescript and spago from source via npm and creating this template. Do you have any recommendations?

Here is the template file I'm working on: https://github.com/waynevanson/void-packages/tree/package/spago_dev/srcpkgs/spago

**Comment by hdgarrood** [comment]
As far as I can tell the reason this is happening is that the package RebindableSyntax and NoImplicitPrelude defined in its cabal file, so that this line uses whichever `fail` is in scope to handle the case where the pattern match fails. It looks to me like `basement` has its own prelude (`Basement.Compat.Base`) and previously `fail` was coming via there, as Basement.Compat.Base has a re-export like this:
```
  Prelude.Monad (..)
```
and `fail` was a member of the `Monad` type class until GHC 8.8. As of GHC 8.8, `fail` is no longer part of `Monad`, so if you are using a newer version of GHC that would explain the error.

Do you know what version of GHC you are using? I think your options are to either downgrade GHC, or choosing a different set of dependencies so that it builds with a version of `basement` which does support whichever version of GHC you're using. I'd strongly recommend finding a way to use the same GHC as the one specified in this project's stack.yaml file, since maintaining a spago package with a newer GHC version than the one used in this project will be an ongoing source of pain for you because of breakages like this (we've seen this repeatedly with the compiler and its homebrew and nix packages).

**Comment by f-f** [comment]
@waynevanson as Harry noted above it's important that you know (and possibly pin) the GHC version you're using to compile this. It looks like you're on GHC 8.8, but Spago currently supports building only with GHC 8.4.

This said, we need to move to 8.8 quite soon (see e.g. #685), so once we do that your problem might go away (to reappear again the next time there's a new GHC version mismatch)

**Comment by waynevanson** [comment]
Very helpful, thank you! I'd like to keep this open until my PR has been accepted by the void team.

**Comment by f-f** [comment]
I'd like to keep the issue tracker as "actionable" as possible, so I wouldn't like to keep an issue open just to track the status of a ticket in another repo - that is, https://github.com/void-linux/void-packages/pull/24723 is a fine place to track itself.
So I'll close this, and note that now that we merged #695 (which upgrades to GHC 8.6.5) things might have improved wrt the compatibility with the GHC that Void is shipping.
